app/Http/Controllers/Backend/CategoryController.php

```diff
@@ -0,0 +1,134 @@
+        $list = new Category;
+        return view('backend.categories.create', compact('cates', 'list'));
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param \Illuminate\Http\Request $request request
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function store(CategoryRequest $request)
+    {
+        $addcate = new Category;
+        $addcate->cate_name = $request->cate_name;
+        $addcate->cate_description = $request->cate_description;
+        $addcate->cate_status = $request->cate_status;
+        $addcate->parent_id = $request->parent_id;
+        if ($request->hasFile('cate_image')) {
+            $fileName = $request->file('cate_image')->getClientOriginalName();
+            $addcate->cate_image = $fileName;
+            $request->file('cate_image')->move(public_path('upload/'), $fileName);
+        }
+        $addcate->save();
+        $request->session()->flash('message', 'Category was add successfully!');
+        return redirect()->route('admin.categories.index');
+    }
+
+    /**
+     * Show the form for editing the specified resource.
+     *
+     * @param int $id id
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function edit($id)
+    {
+        $data = Category::findOrFail($id);
+        $cates = Category::select('id', 'cate_name', 'parent_id')->get()->toArray();
+        $list = new Category;
+        return view('backend.categories.edit', compact('data', 'cates', 'list'));
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param \Illuminate\Http\Request $request CategoryRequest
+     * @param int                      $id      id
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function update(CategoryRequest $request, $id)
+    {
+        $editcate = Category::findOrFail($id);
+        $editcate->cate_name = $request->cate_name;
+        $editcate->cate_description = $request->cate_description;
+        $editcate->cate_status = $request->cate_status;
+        $editcate->parent_id = $request->parent_id;
+        if ($request->hasFile('cate_image')) {
+            if ($editcate->cate_image) {
+                file::delete(public_path('upload/') . $editcate->cate_image);
+            }
+            $fileName = $request->file('cate_image')->getClientOriginalName();
+            $editcate->cate_image = $fileName;
+            $request->file('cate_image')->move(public_path('upload/'), $fileName);
+        }
+        $editcate->save();
+        $request->session()->flash('message', 'Category was update successfully!');
+        return redirect()->route('admin.categories.index');
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param int $id id
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function destroy($id)
+    {
+        $delcate = Category::findOrFail($id);
```

Comment:
[rank S] what's it means? I think you need thinking how to optimise your code and avoid make it like spaghettis